src/lib/storage/table.cpp

```diff
@@ -373,39 +403,114 @@ void Table::create_chunk_index(const std::vector<ColumnID>& column_ids, const st
-      // Ensure that no other PRIMARY KEY is defined
-      Assert(existing_constraint.key_type() == KeyConstraintType::UNIQUE ||
-                 table_key_constraint.key_type() == KeyConstraintType::UNIQUE,
-             "Another primary key already exists for this table.");
+  for (const auto& existing_constraint : _table_key_constraints) {
+    // Ensure that no other PRIMARY KEY is defined.
+    Assert(existing_constraint.key_type() == KeyConstraintType::UNIQUE ||
+               table_key_constraint.key_type() == KeyConstraintType::UNIQUE,
+           "Another primary TableKeyConstraint exists for this table.");
 
-      // Ensure there is only one key constraint per column set.
-      Assert(table_key_constraint.columns() != existing_constraint.columns(),
-             "Another key constraint for the same column set has already been defined.");
-    }
+    // Ensure there is only one key constraint per column set.
+    Assert(table_key_constraint.columns() != existing_constraint.columns(),
+           "Another TableKeyConstraint for the same column set has already been defined.");
+  }
 
-    _table_key_constraints.insert(table_key_constraint);
+  _table_key_constraints.insert(table_key_constraint);
+}
+
+const TableKeyConstraints& Table::soft_key_constraints() const {
+  return _table_key_constraints;
+}
+
+void Table::add_soft_foreign_key_constraint(const ForeignKeyConstraint& foreign_key_constraint) {
+  Assert(_type == TableType::Data, "ForeignKeyConstraints are not tracked for reference tables across the PQP.");
+
+  Assert(&*foreign_key_constraint.foreign_key_table() == this, "ForeignKeyConstraint is added to the wrong table.");
+
+  // Check validity of specified columns.
+  const auto column_count = this->column_count();
+  for (const auto& column_id : foreign_key_constraint.foreign_key_columns()) {
+    Assert(column_id < column_count, "ColumnID out of range.");
+  }
+
+  const auto referenced_table = foreign_key_constraint.primary_key_table();
+  Assert(referenced_table && &*referenced_table != this, "ForeignKeyConstraint must reference another existing table.");
+
+  // Check validity of key columns from other table.
+  const auto referenced_table_column_count = referenced_table->column_count();
+  for (const auto& column_id : foreign_key_constraint.primary_key_columns()) {
+    Assert(column_id < referenced_table_column_count, "ColumnID out of range.");
+  }
+
+  const auto append_lock = acquire_append_mutex();
+  for (const auto& existing_constraint : _foreign_key_constraints) {
+    // Do not allow overlapping foreign key constraints. Though a table may have unlimited inclusion dependencies for
+    // the same columns (and especially the existence of [a] in [x] and [b] in [y] does not mean [a, b] in [x, y]
+    // holds), it is reasonable to assume disjoint (soft) foreign keys for now.
+    Assert(!columns_intersect(existing_constraint.foreign_key_columns(), foreign_key_constraint.foreign_key_columns()),
```

Comment:
You say "overlapping" and use intersect, while there is a function called `columns_overlap`.